### PR TITLE
Dbkz

### PR DESCRIPTION
**Pull request by blowfish880**
An extension to simple BKZ (simple_bkz.py) implementing DBKZ. This seems to work. I think including it in the actual BKZ implementation (bkz.py) can be done similarly, but maybe some refactoring would be desirable for that. For example, I think it would look cleaner, if there was a base class called "lattice reduction" (or so) that implements SVP and DSVP reduction and the derived classes, BKZ and DBKZ (and Slide reduction?) simply string them together in the right order. What do you think?

### CODE REVIEW

File: src/fpylll/contrib/simple_bkz.py [hunk 1]
```diff
@@ -110,3 +110,59 @@ def svp_reduction(self, kappa, block_size):
             self.m.remove_last_row()
 
         return False
+
+class DBKZReduction(BKZReduction):
```
Comment (malb) [inline src/fpylll/contrib/simple_bkz.py]: I personally would prefer this class to have its own file, but won't insist on it.

File: src/fpylll/contrib/simple_bkz.py [hunk 2]
```diff
@@ -110,3 +110,59 @@ def svp_reduction(self, kappa, block_size):
...
+                if not x1:
+                    return pair2
+                c = math.floor(x2/x1)
+                self.m.row_addmul(row2, row1, -c)
+                return euclid((row2, x2 - c*x1), pair1)
+            
+            pairs = list(enumerate(solution, start=kappa))
+            [self.m.negate_row(pair[0]) for pair in pairs if pair[1] < 0]
+            pairs = map(lambda x: (x[0], abs(x[1])), pairs)
+            # GCD should be tree based but for proof of conceot implementation, this will do
```
Comment (malb) [inline src/fpylll/contrib/simple_bkz.py]: conceot -> concept

File: src/fpylll/contrib/simple_bkz.py [hunk 3]
```diff
@@ -110,3 +110,59 @@ def svp_reduction(self, kappa, block_size):
...
+            row, x = reduce(euclid, pairs) 
+            if (x != 1):
```
Comment (malb) [inline src/fpylll/contrib/simple_bkz.py]: No brackets needed

File: src/fpylll/contrib/simple_bkz.py [hunk 4]
```diff
@@ -110,3 +110,59 @@ def svp_reduction(self, kappa, block_size):
...
+        max_dist = 1.0/max_dist
+        expo *= -1.0
+        delta_max_dist = self.lll_obj.delta * max_dist
+        
+        solution, max_dist = Enum.enumerate(self.m, max_dist, expo, kappa, kappa + block_size, None, dual=True)
+        if max_dist >= delta_max_dist:
+            return clean
+        
+        with self.m.row_ops(kappa, kappa+block_size):
+            def euclid(pair1, pair2):
```
Comment (malb) [inline src/fpylll/contrib/simple_bkz.py]: An advantage of putting this on the top level would be that we can test it better.

**Comment by malb** [comment]
How annoying would it be for you to add a test similar to https://github.com/malb/fpylll/blob/master/tests/test_lll.py I can also add a test_bkz.py first, if that'd be helpful.

**Comment by malb** [comment]
Shall we discuss adapting the "full" BKZ in a separate ticket in order not to overload this one?

**Comment by blowfish880** [comment]
I added the changes you suggested. The LLL test file is only testing if all float types are doing the same thing? I can certainly do that for DBKZ. Or did you also want some functionality tests? Maybe it would be easier if you simply wrote a test file for BKZ and I'll structure the one for DBKZ after that. But all in all that might be more relevant once we have a more complete implementation of DBKZ (including local preprocessing etc). As I said, this should also be relatively straight forward. But yes, we should maybe discuss that in a separate ticket.

**Pull request merged**